src/v/datalake/translation/partition_translator.cc

```diff
@@ -0,0 +1,273 @@
+      max_attempts,
+      [this, request = std::move(request)] {
+          return _frontend->local().add_translated_data_files(request.copy());
+      },
+      [this](coordinator::add_translated_data_files_reply result) {
+          return can_continue() && is_retriable(result.errc);
+      });
+    vlog(_logger.trace, "Adding tranlsated data file, result: {}", result);
+    co_return result.errc == coordinator::errc::ok
+      && !(co_await _stm->sync_with_coordinator(
+        last_offset, _term, wait_timeout, _as));
+}
+
+ss::future<std::optional<kafka::offset>>
+partition_translator::reconcile_with_coordinator() {
+    if (_reconcile) {
+        auto request = coordinator::fetch_latest_data_file_request{};
+        request.tp = _partition->ntp().tp;
+        vlog(_logger.trace, "fetch_latest_data_file, request: {}", request);
+        auto resp = co_await _frontend->local().fetch_latest_data_file(request);
+        vlog(_logger.trace, "fetch_latest_data_file, response: {}", resp);
+        if (resp.errc != coordinator::errc::ok) {
+            co_return std::nullopt;
+        }
+        // No file entry signifies the translation was just enabled on the
+        // topic. In such a case we start translation from the local start
+        // of the log. The underlying assumption is that there is a reasonable
+        // retention policy in place that clamps how big the local log is.
+        auto sync_offset = resp.last_added_offset
+                             ? resp.last_added_offset.value()
+                             : kafka::prev_offset(min_offset_for_translation());
+        auto sync_error = co_await _stm->sync_with_coordinator(
+          sync_offset, _term, wait_timeout, _as);
+        if (sync_error) {
+            co_return std::nullopt;
+        }
+    }
+    // Reconciliation is an optimization to avoid wasteful
+    // translation work (particularly on term changes). Worst case
+    // we do translation work on a stale offset range which then gets
+    // rejected on the coordinator checkpoint and we are forced to reconcile
+    // again. Here we try to avoid round trips to the coordinator when
+    // the translator thinks it is not stale.
+    co_return co_await _stm->highest_translated_offset(wait_timeout);
+}
+
+bool partition_translator::can_continue() const {
+    return !_as.abort_requested() && _term == _partition->raft()->term();
+}
+
+ss::future<> partition_translator::do_translate() {
+    while (can_continue()) {
+        co_await ss::sleep_abortable(_jitter.next_duration(), _as);
+        backoff retry_policy{_jitter.base_duration()};
+        auto result = co_await retry_policy.retry(
+          max_attempts,
+          [this] { return do_translate_once(); },
+          [this](translation_success result) {
+              auto retry = can_continue() && result == translation_success::no;
+              if (retry) {
+                  _reconcile = needs_reconciliation::yes;
+              }
```

Comment:
I'm wondering if it makes sense to just always try to sync with the coordinator on failure within do_translate_once -- then we wouldn't need to keep track of _reconcile, and the happy path of do_translate could always rely on the STM